contracts/mapping-token/contracts/v2/Backing.sol

```diff
@@ -0,0 +1,270 @@
+        require(outboundLane != address(0), "cannot find outboundLane to send message");
+        bytes memory newErc20Contract = abi.encodeWithSelector(
+            IMappingTokenFactory.newErc20Contract.selector,
+            address(this),
+            ERC20_TOKEN_TYPE,
+            token,
+            thisChainName,
+            name,
+            symbol,
+            decimals
+        );
+        uint256 fee = IFeeMarket(feeMarket).market_fee();
+        require(msg.value >= fee, "not enough fee to pay");
+        uint64 nonce = IOutboundLane(outboundLane).send_message{value: fee}(remoteMappingTokenFactory, newErc20Contract);
+        uint256 messageId = encodeMessageId(bridgedLanePosition, nonce);
+        registerMessages[messageId] = token;
+        if (msg.value > fee) {
+            (bool sent,) = msg.sender.call{value: msg.value - fee}("");
+            require(sent, "transfer back fee failed");
+        }
+        emit NewErc20TokenRegistered(messageId, bridgedLanePosition, token);
+    }
+
+    /**
+     * @notice lock original token and issuing mapping token from bridged chain
+     * @dev maybe some tokens will take some fee when transfer
+     * @param bridgedLanePosition the bridged lane positon, this issuing message will be delived to this lane position
+     * @param token the original token address
+     * @param recipient the recipient who will receive the issued mapping token
+     * @param amount amount of the locked token
+     */
+    function lockAndRemoteIssuing(
+        uint32 bridgedLanePosition,
+        address token,
+        address recipient,
+        uint256 amount
+    ) external payable whenNotPaused {
+        require(registeredTokens[token], "Backing: the token is not registed");
+
+        uint256 balanceBefore = IERC20(token).balanceOf(address(this));
+        require(IERC20(token).transferFrom(msg.sender, address(this), amount), "transfer tokens failed");
+        uint256 balanceAfter = IERC20(token).balanceOf(address(this));
+        uint256 newAmount = balanceAfter - balanceBefore;
+        address outboundLane = outboundLanes[bridgedLanePosition];
+        require(outboundLane != address(0), "Backing: outboundLane not exist");
+        bytes memory issueMappingToken = abi.encodeWithSelector(
+            IMappingTokenFactory.issueMappingToken.selector,
+            address(this),
+            token,
+            recipient,
+            newAmount
+        );
+        uint256 fee = IFeeMarket(feeMarket).market_fee();
+        require(msg.value >= fee, "not enough fee to pay");
+        uint64 nonce = IOutboundLane(outboundLane).send_message{value: fee}(remoteMappingTokenFactory, issueMappingToken);
+        uint256 messageId = encodeMessageId(bridgedLanePosition, nonce);
+        lockMessages[messageId] = LockedInfo(token, msg.sender, amount);
+        if (msg.value > fee) {
+            (bool sent,) = msg.sender.call{value: msg.value - fee}("");
```

Comment:
I think `payable(msg.sender).transfer` + `nonReentrant` may be more safe, or even better : Use a pattern where the recipient withdraws the money.

Reply:
I find some docs that explain why not use `send` and `transfer`.
Both of them take a hard dependency on gas costs to be safe against reentrancy, right? @hujw77